File: assign1_submit/ProdCons1.java
/* Author: Mina Vu
 * File:   ProdCons1.java
 * About:  1 producer and 1 consumer program
 */
import java.util.LinkedList;
import java.util.Queue;

public class ProdCons1 {
    final static int BUFSIZE = 20;
    final static int NUMITEMS = 100;
    final static Object syncher = new Object();
    static Queue<Integer> queue = null;

    static Runnable producer = () -> {
        System.out.println("Producer starting...");

        for (int i = 1; i <= NUMITEMS; ++i) {
            synchronized (syncher) {
                while (queue.size() == BUFSIZE) {
                    try {
                        syncher.wait();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                queue.add(i);
                syncher.notify();
            }
            System.out.println("Producer added " + i + " (qsz: " + queue.size() + ")");
        }

        System.out.println("Producer ending...");
    };

    static Runnable consumer = () -> {
        System.out.println("Consumer starting...");

        for (int i = 1; i <= NUMITEMS; ++i) {
            int val;
            synchronized (syncher) {
                while (queue.size() == 0) {
                    try {
                        syncher.wait();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                val = queue.remove();
                syncher.notify();
            }
            System.out.println("Consumer rem'd " + val + " (qsz: " + queue.size() + ")");
        }

        System.out.println("Consumer ending...");
    };

    public static void main(String[] args) {
        queue = new LinkedList<>();
        Thread pThread = new Thread(producer);
        Thread cThread = new Thread(consumer);

        try {
            pThread.start();
            cThread.start();
            pThread.join();
            cThread.join();
        } catch (InterruptedException e) {
            System.err.println(e.getMessage());
        }

        System.out.println("Main: all done!");
    }
}

File: assign1_submit/prodcons1.cpp
/* Author: Mina Vu
 * File:   prodcons1.cpp
 * About:  1 producer and 1 consumer program
 */
#include <iostream>
#include <string>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <sched.h>
#include "queue.h"

using namespace std;

int BUFSIZE = 20;
int NUMITEMS = 100;
mutex mtx;
condition_variable cvar;
Queue* queue;

void producer();
void consumer();

// Main
int main(int argc, char **argv) {
  queue = new Queue(BUFSIZE);
  thread prod(producer);
  thread cons(consumer);

  prod.join();
  cons.join();

  cout << "Main: all done!\n";

  return 0;
}

// Producer
void producer() {
  cout << "Producer starting on core " + to_string(sched_getcpu()) + "\n";

  for (int i {1}; i <= NUMITEMS; ++i) {
    unique_lock<mutex> lck(mtx);
    if (queue->isFull()) {
      cvar.wait(lck);
    }
    queue->add(i);
    lck.unlock();
    cvar.notify_one();
    cout << "Producer added " + to_string(i) + " (qsz: " + to_string(queue->size()) + ")\n";
  }

  cout << "Producer ending\n";
}

// Consumer
void consumer() {
  cout << "Consumer starting on core " + to_string(sched_getcpu()) + "\n";

  for (int i {1}; i <= NUMITEMS; ++i) {
    unique_lock<mutex> lck(mtx);
    if (queue->isEmpty()) {
      cvar.wait(lck);
    }
    int val = queue->remove();
    lck.unlock();
    cvar.notify_one();
    cout << "Consumer rem'd " + to_string(val) + " (qsz: " + to_string(queue->size()) + ")\n";
  }

  cout << "Consumer ending\n";
}
